### PR TITLE
Assist webview

### PR DESCRIPTION
**Pull request by abhideep**

### CODE REVIEW

File: src/java/com/sarality/app/view/web/AssistedWebView.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+import android.view.inputmethod.InputConnection;
+import android.webkit.WebView;
+
+import com.sarality.app.view.EditorView;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A WevView that allows an AssistanceKeyboard to be attached to it.
```
Comment (sunayna) [inline src/java/com/sarality/app/view/web/AssistedWebView.java]: typo

File: src/java/com/sarality/app/view/EditorView.java [hunk 2]
```diff
@@ -0,0 +1,18 @@
+package com.sarality.app.view;
+
+import android.view.inputmethod.EditorInfo;
+import android.view.inputmethod.InputConnection;
+
+/**
+ * Interface for a View that has editable elements
```
Comment (sunayna) [inline src/java/com/sarality/app/view/EditorView.java]: the purpose of this interface is not very clear. what editor info is required?

File: src/java/com/sarality/app/keyboard/KeyboardOutputComposer.java [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+  private void handleBackspace() {
+    final int length = outputBuilder.length();
+    if (length > 1) {
+      outputBuilder.delete(length - 1, length);
+      getCurrentInputConnection().setComposingText(outputBuilder, 1);
+    } else if (length > 0) {
+      outputBuilder.setLength(0);
+      getCurrentInputConnection().commitText("", 0);
+    } else {
+      getCurrentInputConnection().sendKeyEvent(new KeyEvent(KeyEvent.ACTION_DOWN, KeyEvent.KEYCODE_DEL));
```
Comment (sunayna) [inline src/java/com/sarality/app/keyboard/KeyboardOutputComposer.java]: why is up and down key event sent along with del?

**Comment by sunayna** [comment]
rest is good

File: src/java/com/sarality/app/view/EditorView.java (hunk 2 above)
Comment (abhideep) [inline src/java/com/sarality/app/view/EditorView.java]: If you look at the Output Composer it needs access to the Input Connection and the EditorOptions to be able to add things to the editable portions of the View. 

Of course this is not needed when you use the Flight Assistant Input Method which has access to the InputCOnnection as well as Editor Options but if you want to use things without requiring the user to change the Keyboard, you need some mechanism to incept the creation of the InputCOnnection. This interface provides that mechanism

File: src/java/com/sarality/app/keyboard/KeyboardOutputComposer.java (hunk 3 above)
Comment (abhideep) [inline src/java/com/sarality/app/keyboard/KeyboardOutputComposer.java]: This takes care of the case where there is no text in the string builder. In such cases we'll just send the delete key to the app and let it handle that as the default case.

**Pull request merged**